src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -419,7 +419,27 @@ ccl::DisplacementMethod displacementMethodFromString( const string &name )
 	return ccl::DisplacementMethod::DISPLACE_TRUE;
 }
 
-IECore::InternedString g_shaderUseMisAttributeName( "cycles:shader:use_mis" );
+std::array<IECore::InternedString, ccl::EMISSION_SAMPLING_NUM> g_emissionSamplingEnumNames = { {
+	"none",
+	"auto",
+	"front",
+	"back",
+	"front_back"
+} };
+ccl::EmissionSampling nameToEmissionSamplingMethodEnum( const IECore::InternedString &name )
+{
+#define MAP_NAME(enumName, enum) if(name == enumName) return enum;
+	MAP_NAME(g_emissionSamplingEnumNames[0], ccl::EMISSION_SAMPLING_NONE);
+	MAP_NAME(g_emissionSamplingEnumNames[1], ccl::EMISSION_SAMPLING_AUTO);
+	MAP_NAME(g_emissionSamplingEnumNames[2], ccl::EMISSION_SAMPLING_FRONT);
+	MAP_NAME(g_emissionSamplingEnumNames[3], ccl::EMISSION_SAMPLING_BACK);
+	MAP_NAME(g_emissionSamplingEnumNames[4], ccl::EMISSION_SAMPLING_FRONT_BACK);
+#undef MAP_NAME
+
+	return ccl::EMISSION_SAMPLING_AUTO;
+}
```

Comment:
I know this is the way things have been done historically in this file, but duplicating all the information from the enums like this is fragile (as the `g_volumeSamplingEnumNames` bug demonstrates). Can't we just use the NodeEnum API itself to do the mapping for us, like we do in `setEnumSocket` in `SocketAlgo.cpp`? Then everything is automatically in sync when Cycles changes the enum...

Reply:
Yeah I was thinking this is probably better and it'll shrink the code a bit.

Reply:
Ok I've got a new `applyShaderEnum` which should do the trick.